lightclient/src/background.rs

```diff
@@ -46,29 +50,160 @@ pub enum FromSubxt {
+    pub async fn subscribe(
+        &self,
+        method: String,
+        params: Option<Box<RawValue>>,
+        unsubscribe_method: String,
+    ) -> SubscriptionResponse {
+        let (tx, rx) = oneshot::channel();
+        self.to_backend
+            .send(Message::Subscription {
+                method,
+                params,
+                unsubscribe_method,
+                sender: tx,
+            })
+            .map_err(|_e| LightClientRpcError::BackgroundTaskDropped)?;
+
+        match rx.await {
+            Err(_e) => Err(LightClientRpcError::BackgroundTaskDropped),
+            Ok(response) => response,
+        }
+    }
+}
+
+/// A background task which runs with [`BackgroundTask::run()`] and manages messages
+/// coming to/from Smoldot.
 #[allow(clippy::type_complexity)]
 pub struct BackgroundTask<TPlatform: PlatformRef, TChain> {
-    /// Smoldot light client implementation that leverages the exposed platform.
-    client: smoldot_light::Client<TPlatform, TChain>,
-    /// Per-chain data.
-    chain_data: HashMap<smoldot_light::ChainId, ChainData>,
+    channels: BackgroundTaskChannels,
+    data: BackgroundTaskData<TPlatform, TChain>,
 }
 
-/// The data that we store for each chain.
-#[derive(Default)]
-struct ChainData {
-    /// Generates an unique monotonically increasing ID for each chain.
+impl<TPlatform: PlatformRef, TChain> BackgroundTask<TPlatform, TChain> {
+    /// Constructs a new [`BackgroundTask`].
+    pub(crate) fn new(
+        client: SharedClient<TPlatform, TChain>,
+        chain_id: smoldot_light::ChainId,
+        from_back: smoldot_light::JsonRpcResponses,
+    ) -> (BackgroundTask<TPlatform, TChain>, BackgroundTaskHandle) {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let bg_task = BackgroundTask {
+            channels: BackgroundTaskChannels {
+                from_front: UnboundedReceiverStream::new(rx),
+                from_back,
+            },
+            data: BackgroundTaskData {
+                client,
+                chain_id,
+                last_request_id: 0,
+                pending_subscriptions: HashMap::new(),
+                requests: HashMap::new(),
+                subscriptions: HashMap::new(),
+            },
+        };
+
+        let bg_handle = BackgroundTaskHandle { to_backend: tx };
+
+        (bg_task, bg_handle)
+    }
+
+    /// Run the background task, which:
+    /// - Forwards messages/subscription requests to Smoldot from the front end.
+    /// - Forwards responses back from Smoldot to the front end.
+    pub async fn run(self) {
+        let chain_id = self.data.chain_id;
+        let mut channels = self.channels;
+        let mut data = self.data;
+
+        loop {
+            tokio::pin! {
+                let from_front_fut = channels.from_front.next().fuse();
+                let from_back_fut = channels.from_back.next().fuse();
+            }
+
+            futures::select! {
```

Comment:
I moved to this I think to avoid allocating somewhere or annoying lifetime things; can't remember exactly now. It just takes two channels as inputs though and so I think should have no cancellation concerns (this is the usual footgun with `select!`

Reply:
Looks good, btw you asked me exactly this question "What is the usual usual footgun with select!" when interviewing me as a job applicant about a year ago.

Reply:
I need to make a note somewhere because if I move to doing a code review style task then `future::select!` is def a candidate to put somewhere :D

Reply:
you are good, streams are cancel-safe.

I like tokio::select better because it doesn't require pinning the futures

Reply:
I normally prefer the function call version to avoid some footguns but iirc I had an issue that made it hard to use it nicely; I'll have to try again at some point!